APE6.rst

```diff
@@ -0,0 +1,499 @@
+In simple cases the column definitions serialize on a single line which
+makes for a compact and useful representation.
+
+Example
+^^^^^^^^^^
+
+A couple of quick examples using the reference Python implementation
+will put this in context.  First let's create a table and give it some
+custom attributes::
+
+  >>> from astropy.table import Table
+  >>> t = Table([[1, 4], [2, 3]], names=['a', 'b'])
+  >>> t['a'].unit = 'm/s'
+  >>> t['a'].format = '%03d'
+  >>> t['b'].description = 'This is column b'
+  >>> t['b'].unit = 'km'
+  >>> print(t)
+    a    b 
+  m / s  km
+  ----- ---
+    001   2
+    004   3
+
+Now we write this to a file using the ECSV format and print it::
+
+  >>> t.write('example.ecsv', format='ascii.ecsv')
+  >>> cat example.ecsv
+  # %ECSV 0.9
+  # ---
+  # datatype:
+  # - {name: a, unit: m / s, datatype: int64, format: '%03d'}
+  # - {name: b, unit: km, datatype: int64, description: This is column b}
+  a b
+  001 2
+  004 3
+
+We see that header starts with a header line and YAML block marker to identify
+the format and the beginning of the data block.  After that comes the YAML data
+structure with column definitions.  Finally the column names and data values
+are included in CSV format with a space delimiter.
+
+Now we can read back the table and see that it has survived the round-trip
+to a text file::
+
+  >>> t2 = Table.read('example.ecsv', format='ascii.ecsv')
+  >>> print(t2)
+    a    b 
+  m / s  km
+  ----- ---
+    001   2
+    004   3
+
+To illustrate the full features of ECSV we create a table with meta data
+(keywords and comments) as well as more detailed column meta data::
+
+  >>> from astropy.utils import OrderedDict
+  >>> t = Table([[1, 4], [2, 3]], names=['a', 'b'])
+  >>> t.meta['keywords'] = OrderedDict([('z_key1', 'val1'), ('a_key2', 'val2')])
+  >>> t.meta['comments'] = ['Comment 1', 'Comment 2', 'Comment 3']
+  >>> t['a'].unit = u.m / u.s
+  >>> t['a'].format = '%5.2f'
+  >>> t['a'].description = 'Column A'
+  >>> t['b'].meta = dict(column_meta={'a':1, 'b': 2})
+
+Now we write the table to standard out::
+
+  >>> import sys
+  >>> t.write(sys.stdout, format='ascii.ecsv')
+  # %ECSV 0.9
+  # ---
+  # datatype:
+  # - {name: a, unit: m / s, type: int64, format: '%5.2f', description: Column A}
+  # - name: b
+  #   datatype: int64
+  #   meta:
+  #     column_meta: {a: 1, b: 2}
+  # meta: !!omap
+  # - keywords: !!omap
+  #   - {z_key1: val1}
+  #   - {a_key2: val2}
+  # - comments: [Comment 1, Comment 2, Comment 3]
+  a b
+  1.00 2
+  4.00 3
+
+In this case there are Ordered dictionary elements which must be preserved
+during serialization.  This is done via the ``!!omap`` element tag.  The
+reference Python implementation in Astropy uses custom Loader and Dumper
+classes to allow serialization to and from the Python ``OrderedDict`` class.
+Other implementations must likewise use an ordered mapping when reading and the
+``!!omap`` tag when writing for ordered mappings in the data structure.
```

Comment:
What do you usually say as your first example of why it's needed?  For me it is crucial for human readability and file round-tripping, but I don't know of examples where there is a real contextual dependence on keyword order.